### PR TITLE
Type inference updates

### PR DESCRIPTION
**Pull request by TPei**

### CODE REVIEW

File: src/aasm/state.cr [hunk 1]
```diff
@@ -1,6 +1,8 @@
 struct AASM::State
 
   getter :enter, :guard
+  @enter : ((->) | Nil)
```
Comment (TPei) [inline src/aasm/state.cr:4]: `@enter : ((-> Any) | Nil)` is basically what we need

Comment (TPei) [inline src/aasm/state.cr:4]: `@enter : ((-> Object) | Nil)`
=> `can't use Object as proc return type yet, use a more specific type`

**Comment by veelenga** [comment]
Fixed by 9256c2fa0b19e0b59bbd4e199dfaac4087ced82a on master.

I temporary fixed that by setting a type of enter block to `(-> Nil)?`. That means you will need to explicitly return `nil`. For example:

``` crystal
aasm.state :active,  enter: -> { @count += 1; nil }
```

That is going to be improved in coming Crystal release: [crystal-lang#2672](https://github.com/crystal-lang/crystal/issues/2672) and seems we will not need to return the type explicitly.

**Comment by veelenga** [comment]
Be free to issue new PR with changes compiled vs crystal master branch, if you need it immediately...

**Comment by veelenga** [comment]
@TPei since b14bc68d930761d49ebaa6c83002f8580cd4f044 it is possible not to return nil explicitly. It works with Crystal's latest release and master branch. Cheers ;)

**Pull request closed without merging**